Create container-build.yml

aryanbhosale opened the pull request:
# Pull Request

## Description

github workflow for creating a docker image for the api 

## Checklist:

- [ ] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked my code and corrected any misspellings

aryanbhosale commented:
hey @peterdudfield , #197 removes requirements.txt and updates pyproject.toml, due to which the RUN pip install -r requirements.txt line had to be removed from the Dockerfile, but it's still throwing an error at RUN pip install -e . 
any suggestions?

aryanbhosale commented:
> hey @peterdudfield , #197 removes requirements.txt and updates pyproject.toml, due to which the RUN pip install -r requirements.txt line had to be removed from the Dockerfile, but it's still throwing an error at RUN pip install -e . any suggestions?

was [this](https://github.com/openclimatefix/open-source-quartz-solar-forecast/pull/192#pullrequestreview-2290428330) tested?

peterdudfield commented:
Try on the latest branch on main? Also what is the error?

aryanbhosale commented:
> Try on the latest branch on main? Also what is the error?

Dockerfile throws the issue at RUN pip install -e . when trying to install pyproject.toml, ill check on my local system without docker once

aryanbhosale commented:
> Try on the latest branch on main? Also what is the error?

it works on my local machine, but on docker it can't find quartz_solar_forecast

aryanbhosale commented:
Okayy so the issue is that pip install -e . works well in my local machine, but when it tries to build the container from the dockerfile, that same command fails, here's one not so helpful discussion i found https://stackoverflow.com/questions/78044103/pip-install-in-editable-mode-from-a-pyproject-in-a-docker-container-modulenotfo
any help on this would be appreciated

peterdudfield commented on Dockerfile:
```diff
@@ -4,20 +4,14 @@ FROM python:3.11-slim
-
-# Install any needed packages specified in requirements.txt
-RUN pip install --no-cache-dir -r requirements.txt
-
```
you need to copy the `pyproject.toml` file over

aryanbhosale replied:
> you need to copy the `pyproject.toml` file over

Aren't we copying all the files in the directory anyway?

peterdudfield replied:
Not currently. In general we try to only copy the files we need. 
`COPY pyproject.toml . ` should do it.

peterdudfield replied:
Ah see, below we do, but we copy them to `app`

aryanbhosale replied:
> Ah see, below we do, but we copy them to `app`

So simply COPY pyproject.toml . Should work?

peterdudfield replied:
I think so, worth a try.

peterdudfield replied:
you can run `docker build` locally to see if it works

aryanbhosale replied:
> you can run `docker build` locally to see if it works

You mean docker compose build?

aryanbhosale replied:
/tmp/pip-build-env-eqosh4xk/overlay/lib/python3.11/site-packages/setuptools/command/build_py.py:218: _Warning: Package 'numcodecs.tests.package_with_entrypoint' is absent from the `packages` configuration.
#16 259.7       !!
#16 259.7       
#16 259.7               ********************************************************************************
#16 259.7               ############################
#16 259.7               # Package would be ignored #
#16 259.7               ############################
#16 259.7               Python recognizes 'numcodecs.tests.package_with_entrypoint' as an importable package[^1],
#16 259.7               but it is absent from setuptools' `packages` configuration.
#16 259.7       
#16 259.7               This leads to an ambiguous overall configuration. If you want to distribute this
#16 259.7               package, please make sure that 'numcodecs.tests.package_with_entrypoint' is explicitly added
#16 259.7               to the `packages` configuration field.
#16 259.7       
#16 259.7               Alternatively, you can also rely on setuptools' discovery methods
#16 259.7               (for example by using `find_namespace_packages(...)`/`find_namespace:`
#16 259.7               instead of `find_packages(...)`/`find:`).
#16 259.7       
#16 259.7               You can read more about "package discovery" on setuptools documentation page:
#16 259.7       
#16 259.7               - https://setuptools.pypa.io/en/latest/userguide/package_discovery.html
#16 259.7       
#16 259.7               If you don't want 'numcodecs.tests.package_with_entrypoint' to be distributed and are
#16 259.7               already explicitly excluding 'numcodecs.tests.package_with_entrypoint' via
#16 259.7               `find_namespace_packages(...)/find_namespace` or `find_packages(...)/find`,
#16 259.7               you can try to use `exclude_package_data`, or `include-package-data=False` in
#16 259.7               combination with a more fine grained `package-data` configuration.
#16 259.7       
#16 259.7               You can read more about "package data files" on setuptools documentation page:
#16 259.7       
#16 259.7               - https://setuptools.pypa.io/en/latest/userguide/datafiles.html


this is the issue @peterdudfield

aryanbhosale replied:
> #18 2.100 configuration error: `tool.setuptools.packages` must be valid exactly by one definition (0 matches found):
> 
> this is the error, it means the packages section of pyproject.toml isnt correctly set

this seems to be the point of failure:
Failed to build numcodecs h5py
220.5 ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (numcodecs, h5py)

seems that h5py requires some additional dependencies , trying that

aryanbhosale replied:
okay i tried to do a workaround here, tried installing h5py which was creating issues with numcodecs as seen in the error logs, 
but i see this:
Dockerfile:14
--------------------
  12 |     
  13 |     # Install h5py with no-binary flag
  14 | >>> RUN pip install --no-binary h5py h5py
  15 |     
  16 |     # Copy the pyproject.toml file
--------------------
ERROR: failed to solve: process "/bin/sh -c pip install --no-binary h5py h5py" did not complete successfully: exit code: 1

apparently everything fails at the first RUN pip install xyz(whatever dependency)

peterdudfield replied:
do you need `pip install --no-binary=h5py h5py`, I admit this is a weird problem though

aryanbhosale replied:
> do you need `pip install --no-binary=h5py h5py`, I admit this is a weird problem though

Not really no, was just trying to see if installing it explicitly fixes the h5py, numcodec issue we see on docker, but rather it led us to another issue that the first pip install whatever always fails

aryanbhosale commented:
hey @peterdudfield , ive removed all the extra changes like adding the frontend to docker, even then i get the same error
locally i get:
failed to solve: process "/bin/sh -c pip install --no-cache-dir build &&     pip install --no-cache-dir ." did not complete successfully: exit code: 2    

and on github i get:
#15 266.2 [numcodecs] command 'gcc' failed: No such file or directory
274.5 ERROR: Could not build wheels for numcodecs, ndindex, which is required to install pyproject.toml-based projects
1687

peterdudfield commented:
Hmm, yea, you might have to install conda abd then install it. See uk-pvnet-app for an example - I think

aryanbhosale commented:
> Hmm, yea, you might have to install conda abd then install it. See uk-pvnet-app for an example - I think

trying this

aryanbhosale commented:
> Hmm, yea, you might have to install conda abd then install it. See uk-pvnet-app for an example - I think

same error :/  
30 | >>> RUN pip install --no-cache-dir -e .

ERROR: failed to solve: process "conda run -n myenv /bin/bash -c pip install --no-cache-dir -e ." did not complete successfully: exit code: 1

aryanbhosale commented:
> > Hmm, yea, you might have to install conda abd then install it. See uk-pvnet-app for an example - I think
> 
> same error :/ 30 | >>> RUN pip install --no-cache-dir -e .
> 
> ERROR: failed to solve: process "conda run -n myenv /bin/bash -c pip install --no-cache-dir -e ." did not complete successfully: exit code: 1

@peterdudfield it just refuses to install .(the current dir):
ERROR: failed to solve: process "conda run -n myenv /bin/bash -c pip install --no-cache-dir -e ." did not complete successfully: exit code: 1

aryanbhosale commented:
Hey @peterdudfield @Sukh-P , fixed this
The issue was that the version of numcodecs specified in the new pyproject.toml file requires rust, so i just added and ran that to the dockerfile and made it work, guess this can be merged

peterdudfield commented:
Ill close this now, as we have some CI to build and push docker images
